src/util/tokenpipe.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+#include <util/tokenpipe.h>
+
+#ifndef WIN32
+
+#include <errno.h>
+#include <unistd.h>
+
+TokenPipe::TokenPipe()
+{
+#if HAVE_O_CLOEXEC
+    if (pipe2(m_fds, O_CLOEXEC) != 0) {
+        return;
```

Comment:
I was first thinking of using the rust trick: to have a private constructor, and a static public factory function on the class that returns an `std::optional<Pipe>`. But it would complicate some things. E.g. currently this type is not movable at all, it felt like overkill to implement that just for initialization, but maybe it's worth it if we can make it a standard idiom.